### PR TITLE
[OPT] [DEPRECATED] Deepsparse Engine Support

### PR DESCRIPTION
**Pull request by dbogunowicz**

### CODE REVIEW

File: src/deepsparse/pipelines/text_generation.py [hunk 1]
```diff
@@ -230,6 +232,7 @@ def prompt_inference(
                     run_tokens, kv_cache
                 )
         else:
+            print('second')
```
Comment (dbogunowicz) [inline src/deepsparse/pipelines/text_generation.py:235]:
```suggestion
```

File: src/deepsparse/pipelines/text_generation.py [hunk 2]
```diff
@@ -221,6 +222,7 @@ def prompt_inference(
         new_token = None
 
         if len(tokens) / float(self.sequence_length) < self.prompt_batch_threshold:
+            print('first')
```
Comment (dbogunowicz) [inline src/deepsparse/pipelines/text_generation.py:225]:
```suggestion
```

File: src/deepsparse/pipeline.py [hunk 3]
```diff
@@ -263,7 +263,7 @@ def __call__(self, *args, **kwargs) -> BaseModel:
         batches = self.split_engine_inputs(engine_inputs, self._batch_size)
 
         # submit split batches to engine threadpool
-        batch_outputs = list(self.executor.map(self.engine_forward, batches))
+        batch_outputs = [self.engine_forward(x) for x in batches]
```
Comment (dbogunowicz) [inline src/deepsparse/pipeline.py:266]:
```suggestion
        batch_outputs = list(self.executor.map(self.engine_forward, batches))
```

File: src/deepsparse/pipelines/text_generation.py [hunk 4]
```diff
@@ -407,6 +397,24 @@ def overwrite_onnx_model_inputs(
...
         for external_input in external_inputs:
-            external_input.type.tensor_type.shape.dim[0].dim_value = batch_size
+            if external_input.name == "input_ids":
+                external_input.type.tensor_type.shape.dim[0].dim_value = batch_size
+                external_input.type.tensor_type.shape.dim[1].dim_value = sequence_length
+            elif external_input.name == "attention_mask":
+                external_input.type.tensor_type.shape.dim[0].dim_value = batch_size
+                external_input.type.tensor_type.shape.dim[1].dim_value = sequence_length
+            elif external_input.name.startswith("past_key_values"):
+                external_input.type.tensor_type.shape.dim[0].dim_value = 16  # num heads
```
Comment (bfineran) [inline src/deepsparse/pipelines/text_generation.py:407]: you can get this from `config.num_attention_heads`

File: src/deepsparse/pipelines/text_generation.py [hunk 5]
```diff
@@ -0,0 +1,547 @@
...
+
+from deepsparse import Pipeline
+from deepsparse.transformers.helpers import (
+    get_onnx_path_and_configs,
+    overwrite_transformer_onnx_model_inputs,
+)
+from deepsparse.transformers.pipelines import TransformersPipeline
+
+
+OPT_CACHE_HIDDEN_DIM = 64
```
Comment (dbogunowicz) [inline src/deepsparse/pipelines/text_generation.py:31]: I could not find an elegant way to access this variable (not in the self.config variable sadly)

File: src/deepsparse/pipelines/text_generation.py [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+            - the kv cache that was populated during the inference
+        """
+        new_token = tokens[-1]
+
+        # padding is added to left, so attention mask is 1s from the
+        # right up to the number of total tokens (prompt + generated)
+        attention_mask = numpy.zeros((1, self.sequence_length), dtype=numpy.int64)
+        num_tokens_running = min(len(tokens), self.sequence_length)  # cap by seq len
+        attention_mask[:, -num_tokens_running:] = 1
+        breakpoint()
```
Comment (dbogunowicz) [inline src/deepsparse/pipelines/text_generation.py:289]:
```suggestion
```

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+__all__ = ["TextGenerationPipeline"]
+
+
+def softmax(x: numpy.ndarray) -> numpy.ndarray:
+    """
+    Compute softmax values for x
+    :param x: input array
+    :return: softmax values
+    """
+    return numpy.exp(x) / numpy.sum(numpy.exp(x), axis=0)
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:44]: needs epsilon

**Pull request closed without merging**